### ISSUE TITLE
Put structural type refinements with type members under a language flag.

### ISSUE THREAD

**Issue by odersky** [lampepfl/dotty#7636]
If we have refinement types such as
```scala
C { type T = U}
{ type T }
```
it should be checked that the type member `T` does refine a type member in the parent type of the refinement (i.e. `C` or `Any` above). If it is a new type that does not refine anything, this should only be allowed under a language import such as
```scala
import language:newTypesInRefinements
```
### Why Put It Under a Flag?

1. Neither `nsc` nor `dotc` are really made for handling it correctly. It's an afterthought in both of them, and a constant source of bugs. This is because normal code usually does not exercise the feature. 

2. It's prone to hide mistakes. For instance, I detected in `dotc`
```scala
lastSymbol.asInstanceOf[Designator{ type ThisName = self.ThisName }]
```
Type `Designator` does not have a member `ThisName`, so the type refinement is just confusing. 

3. It's prone to abuse, an open invitation to obscure type hackery. Often, it would be perfectly possible to put the types as schemas into the parent class of a refinement, and this would help documentation and the type checker. But since it's shorter there's a temptation to just leave it out. 

### Why Keep the Feature?

Basically, because DOT allows it. However, DOT does not talk about decidability, so it's fair to restrict the feature if decidability cannot be guaranteed.

**Comment by Katrix** [comment]
Will this flag be needed at definition site, or use site?

**Comment by odersky** [comment]
> Will this flag be needed at definition site, or use site?

I don't understand what that means in this context.

**Comment by Katrix** [comment]
`foo.scala`
```scala
type Const[A] = {
  type λ[_] = A
}
```

`bar.scala`
```scala
trait MyService[F[_]] {
  ...
}

def dummy: MyService[Const[Unit]#λ] = ???
```

Will this produce a warning in foo.scala, bar.scala, or both?

**Comment by odersky** [comment]
It would be in `foo.scala`.

**Comment by soronpo** [comment]
Does this mean the infamous `Aux` pattern will now require this flag? (I realize it's probably unnecessary now in dotty)

**Comment by julienrf** [comment]
> Does this mean the infamous `Aux` pattern will now require this flag?

No, it shouldn’t, because in this case the refinement specializes an existing member of the parent type:

> it should be checked that the type member T does refine a type member in the parent type of the refinement

**Comment by odersky** [comment]
I found in our code base and community build there was only a single instance where we used new type members in structural types: Class `deriving.Mirror` missed two any-kinded type members that were used in refinements. It was easily fixed, and the fix made the code much better documented. This backs up the hypothesis that such type members are not idiomatic Scala code,
and strengthens the argument that they should be put under a flag.

**Comment by milessabin** [comment]
I'm afraid this is used a lot in Scala 2 ... I think an audit of the Scala 2 community build is needed to get a true sense of the impact of this change.

**Comment by Katrix** [comment]
Would it maybe be possible to relax the restriction to not apply to `Any` in Scala 2 compat? I think that would rule out a lot of cases of this in Scala 2

**Comment by neko-kai** [comment]
@milessabin 
> I'm afraid this is used a lot in Scala 2 ... I think an audit of the Scala 2 community build is needed to get a true sense of  the impact of this change.

On the contrary, I think it's probably not very much in existing code... Even when I need an unsound or a self-contradictive type, usually the type member also exists in parent trait, so it doesn't fall under this class. New type members also do not expand implicit search scope in Scala 2 – that would be done by tagging – that adds a new type _parameter_, not type member.

**Comment by odersky** [comment]
> Would it maybe be possible to relax the restriction to not apply to Any in Scala 2 compat? I think that would rule out a lot of cases of this in Scala 2

It's a language feature. So you can simply ignore the feature warning of you don't want to change your imports or build.

**Comment by Katrix** [comment]
Ah, I interpreted `this should only be allowed under a language import such as` as it would be an error otherwise, similar to defining macros without the import.

**Comment by odersky** [comment]
> I'm afraid this is used a lot in Scala 2 ... I think an audit of the Scala 2 community build is needed to get a true sense of the impact of this change.

The problem is that neither `nsc` nor `dotc` supports these constructs well; it's a constant source of bugs. And at least in `dotc` I don't see a way or a reason to change that. So I believe it's better to make that limitation clear at the outset with a language feature.

One reason why we needed the feature previously would have been kind polymorphism. Without `AnyKind`, there's no way to specify a schema for type members that can be instantiated with different kinds. Incidentally, the single case where this showed up in our code base and community build was of that nature: Two any-kinded type members in `Mirror` were missing.

**Comment by neko-kai** [comment]
@odersky 
> One reason why we needed the feature previously would have been kind polymorphism. 

I do use this for kind polymorphism in Scala 2, e.g. to create a type tag for a type of arbitrary kind a macro generates searches for that type applied to type member parameters:

```scala
class MyType[A, B, C]
implicitly[HKTag[{ type Arg[A, B, C] = MyType[A, B, C] }]]
```

**Comment by djspiewak** [comment]
> It's prone to abuse, an open invitation to obscure type hackery. Often, it would be perfectly possible to put the types as schemas into the parent class of a refinement, and this would help documentation and the type checker. But since it's shorter there's a temptation to just leave it out.

I think this is the primary case where this is used today. Obviously, the type lambda example given on this thread is far and away the most common scenario, but kind-projector makes this less relevant in Scala 2, and Dotty itself provides a saner mechanism for type lambdas as part of the language, so I don't see that as a concern going forward. I've seen tricky things done with the interaction between `AnyVal` and unrestricted type refinements, but none of them are things I actually care about.

Kind polymorphism is probably the best overarching term for where this technique has meaningful applications *today*, as @neko-kai noted. In most cases that I've encountered though, the need for kind polymorphic (or at least, *shape* polymorphic) types is bounded by some final form, and this final form can be represented in a parent type. [An example](https://github.com/djspiewak/skolems/blob/master/src/main/scala/skolems/Forall.scala#L21-L28).

So tldr, outside of type lambdas, I don't think this feature sees *too* much adoption today, but it would be interesting to see a more rigorous survey.

**Comment by odersky** [comment]
I think we'll not go ahead with this, after all.
